api/envoy/config/listener/v3/listener_components.proto

```diff
@@ -109,6 +110,163 @@ message FilterChainMatch {
     EXTERNAL = 2;
   }
 
+  // Matches filter chains by the destination port.
+  message DestinationPortInput {
+  }
+
+  // Matches filter chains by the destination IP address.
+  message DestinationIPInput {
+  }
+
+  // Matches filter chains by the source IP address.
+  message SourceIPInput {
+  }
+
+  // Matches filter chains by the directly connected source IP address (this
+  // will only be different from the source IP address when using a listener
+  // filter that overrides the source address, such as the :ref:`Proxy Protocol
+  // listener filter <config_listener_filters_proxy_protocol>`).
+  message DirectSourceIPInput {
+  }
+
+  // Matches filter chains by the source port.
+  message SourcePortInput {
+  }
+
+  // Matches filter chains by the connection source IP type.
+  // Specifies the connection source IP match type. The values include:
+  //
+  // * ``any`` - any connection source,
+  // * ``local`` - matches a connection originating from the same host,
+  // * ``external`` - matches a connection originating from a different host
+  message SourceTypeInput {
+  }
+
+  // Matches filter chains by the server names of a new connection, when detected by one of the listener filters.
+  message SNIInput {
+  }
+
+  // Matches filter chains by the transport protocol.
+  // This value will be compared against the transport protocol of a new connection, when
+  // it's detected by one of the listener filters.
+  //
+  // Suggested values include:
+  //
+  // * ``raw_buffer`` - default, used when no transport protocol is detected,
+  // * ``tls`` - set by :ref:`envoy.filters.listener.tls_inspector <config_listener_filters_tls_inspector>`
+  //   when TLS protocol is detected.
+  message TransportProtocolInput {
+  }
+
+  // Matches filter chains by the application protocol (e.g. ALPN for TLS protocol).
+  // This values will be compared against the application
+  // protocols of a new connection, when detected by one of the listener filters.
+  //
+  // Suggested values include:
+  //
+  // * ``http/1.1`` - set by :ref:`envoy.filters.listener.tls_inspector
+  //   <config_listener_filters_tls_inspector>`,
+  // * ``h2`` - set by :ref:`envoy.filters.listener.tls_inspector <config_listener_filters_tls_inspector>`
+  //
+  // .. attention::
+  //
+  //   Currently, only :ref:`TLS Inspector <config_listener_filters_tls_inspector>` provides
+  //   application protocol detection based on the requested
+  //   `ALPN <https://en.wikipedia.org/wiki/Application-Layer_Protocol_Negotiation>`_ values.
+  //
+  //   However, the use of ALPN is pretty much limited to the HTTP/2 traffic on the Internet,
+  //   and matching on values other than ``h2`` is going to lead to a lot of false negatives,
+  //   unless all connecting clients are known to use ALPN.
+  message ApplicationProtocolInput {
+  }
+
+  // Matches a specific port against a set of possibly overlapping port ranges.
+  message PortMatcher {
```

Comment:
These three matchers seem very useful and may also want to be used in places other than the Listener.  Maybe move each one to its own file in https://github.com/cncf/xds/tree/main/xds/type/matcher/v3?

Reply:
Agree that we can move them, but can we keep this here for now to ease the review? Otherwise, we have to do multi-repo PR which is pain. We can move this out once everything settles.

Reply:
Okay.  For now, please add a comment of the form "FIXME BEFORE MERGING: MOVE THESE TO cncf/xds REPO!" to make sure that we don't forget to do this before we merge.

Reply:
Done.